### PR TITLE
Incompressible flow example makeover

### PR DESCRIPTION
**Pull request by blaisb**
# Description of the problem

Now that a nice version of the documentation exists, it is time to migrate the examples slowly but surely. I decided to start with one of the simpler one and to flesh it out into something more interesting!

Additionally, some default values for the linear solver were not appropriate.

# Description of the solution

- Fixed the non-appropriate default values for the linear solver (this generated a lot of changes in the test output, I am sorry about that).

The 2d lid driven cavity example has been fully reworked. It now includes:

- Comparison with results from the literature
- Results at much higher Reynolds number
- A more didactic explanation of the features in Lethe.

# How Has This Been Tested?

- Compiles well on my on machine and on the CI.

# Future changes

- The next example I will migrate with be the Taylor-Couette one. Understanding the parameter file will also be a priority.

# Comments

This took some time, but I had an amazing time doing it.

### CODE REVIEW

**Comment by blaisb** [comment]
This is an archive of the doc. You just need to open the index.html in your browser and naviguate to the example.

[build.tar.gz](https://github.com/lethe-cfd/lethe/files/7798739/build.tar.gz)

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (technophil98) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:17]: Maybe use inline math for Re?
```suggestion
- Base case (:math:`R_e=400`) : ``/examples/incompressible_flows/cavity.prm``
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (technophil98) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:31]: Inline math for u = 1 m/s?

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (technophil98) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:18]: Path could be a link to the mentioned file on GitHub

**Review by technophil98 — COMMENTED** [review]
I may not understand everything, but doc-wise, it's great work! I left a couple of suggestions for your consideration.

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (blaisb) [reply doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:18]: Good question. I'm not sure which is better here. I guess specifying the path will enable users to find quickly the example they are looking for and run it. Pointing to the github repo would have a similar impact, but would be a bit more indirect I guess.

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (technophil98) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:73]: Double backtick for monospace

```suggestion
The last parameter specifies the ``initial refinement`` of the grid. Most deal.ii grid generators contain a minimal number of cells. For example, the *hyper_cube* mesh is made of a single cell. Indicating an ``initial refinement=6`` implies that the initial mesh is refined 6 times. In 2D, each cell is divided by 4 per refinement. Consequently, the final grid is made of :math:`2^{(2\cdot6)}=4096` cells.
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (technophil98) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:109]: Double backtick for monospace
```suggestion
First, the ``number`` of boundary conditions to be applied must be specified. For each boundary condition, the ``id`` of the boundary as well as its ``type`` must be specified. The left (``0``), right (``1``) and bottom (``2``) walls are static and, consequently, a ``noslip`` boundary condition can be used. This boundary condition imposes :math:`\mathbf{u} = [0,0]^T`. For the top wall, we use the ``function`` boundary type. This type of boundary condition allows us to define the value of the velocity components using ``Function expression``. We set :math:`u=1` and :math:`v=0`. Note that the ``Function expression`` supports writing complex mathematical expression which may depend on the spatial coordinates (:math:`x,y,z`) and on time.
```

**Review by shahabgol — APPROVED** [review]

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:24]:
```suggestion
The lid-driven cavity is a classical fluid dynamics problem . It consists in the flow in an enclosed cavity, where one of the walls (in this example, the top wall) is put in motion at a constant tangential velocity. This case is often used to benchmark the capacity of CFD software because of the simplicity of its geometry and the availability of multiple reference results over a large range of Reynolds numbers. In this example, we investigate this problem in 2D, but simulating this case in 3D would require only a few changes. The geometry and boundary conditions are illustrated in the following figure:
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:33]:
```suggestion
We first investigate this case at a Reynolds number of 400 for which a steady-state solution can be easily obtained. Later, we will increase this Reynolds number to 7500 and leverage some of the advanced functionalities of Lethe to reach a steady-state solution. 
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:42]:
```suggestion
Parameter files are made of subsections which describe a portion of the simulation (e.g. mesh generation, physical properties, simulation control). In parameter files, lines starting with ``#`` are comments. Parameters are set using the following syntax:
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:67]:
```suggestion
The ``type`` specifies the mesh format used. At the moment, Lethe supports two mesh formats: ``dealii`` and ``gmsh``. ``dealii`` meshes are in-situ generated meshes for simple geometries. The type of grid generated is specified by the ``grid type`` parameters and this grid is parametrized by it's ``grid arguments``. We refer to the documentation of the deal.ii `GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html>`_ for a detailed explanation of the available grids. 
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:69]:
```suggestion
Since the lid-driven cavity problem domain is a square, we use the *hyper_cube* ``grid_type``. The arguments of this grid type are the position of the bottom left corner, the position of the top right corner and the option to colorize the boundaries in order to give each of them a unique ID. The IDs will be used to set the boundary conditions on specific parts of the boundary of the domain. The ID given to each face was given in the graphical description of the case. If ``colorize`` option were set to false, all boundaries would have been given the ID ``0``.
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:71]:
```suggestion
It is a bit surprising that the position of the bottom left and the top right corner are specified by a single value. Since the geometry is a square, the position of the corner is specified using a single number, assuming that this identifies both the x and y value associated with that point. Other grid generators, such as the ``hyper_rectangle``, allow for more flexibility.
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:79]:
```suggestion
The ``boundary conditions`` subsection establishes the constraints on different parts of the domain:
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst]:
```suggestion
First, the ``number`` of boundary conditions to be applied must be specified. For each boundary condition, the ``id`` of the boundary as well as its ``type`` must be specified. The left (``0``), right (``1``) and bottom (``2``) walls are static and, consequently, a ``noslip`` boundary condition can be used. This boundary condition imposes :math:`\mathbf{u} = [0,0]^T`. For the top wall, we use the `function` boundary type. This type of boundary condition allows us to define the value of the velocity components using `Function expression`. We set :math:`u=1` and :math:`v=0`. Note that the `Function expression` supports writing complex mathematical expressions which may depend on the spatial coordinates (:math:`x,y,z`) and on time.
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:143]:
```suggestion
    An alternative would be to use the ``gd_navier_stokes_2d`` solver; for `LBB <https://en.wikipedia.org/wiki/Ladyzhenskaya%E2%80%93Babu%C5%A1ka%E2%80%93Brezzi_condition>`_ stable elements must be used (e.g. Qn-Q(n-1)). Only the stabilized solver supports the use of equal order elements. 
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:157]:
```suggestion
The ``verbosity`` option specifies if details about the non-linear solver steps (residual value and iteration number) will be printed out to the terminal. By setting it to ``verbose``, this information is printed out, whereas ``quiet`` would mute all outputs of the non-linear solver. We recommend to always set ``verbosity=verbose`` in order to monitor possible non-convergence of the solver.
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:162]:
```suggestion
Each non-linear solver step requires the solution of a linear system of equations. Lethe has multiple options to parametrize its linear solvers, but here, we only enable its verbosity to monitor the number of iteration per time step and use the default parameters for the rest. 
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:165]:
```suggestion
    A good thing to remember is that, generally, linear solver parameters control the robustness of the simulation.
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:213]:
```suggestion
We note that the agreement is perfect. This is not surprising, especially considering that these results were obtained at a relatively low Reynolds number.
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:216]:
```suggestion
    The vtu files generated by Lethe are compressed archives. Consequently, they cannot be postprocessed directly. Although they can be easily post-processed using Paraview, it is sometimes necessary to be able to work with the raw data. The python library `PyVista <https://www.pyvista.org/>`_  allows us to do this.
```

**Comment by blaisb** [comment]
Damn @technophil98  you have the eye of the tiger.

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:301]:
```suggestion
- **High-order methods:** Lethe supports higher order interpolation. This can yield much better results with an equal number of degrees of freedom than traditional second-order (Q1-Q1) methods, especially at higher Reynolds numbers. 
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:303]:
```suggestion
- **Dynamic mesh adaptation:** Lethe supports dynamic mesh adaptation. Running this case with dynamic mesh adaptation could potentially yield better results.
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:300]:
```suggestion
- **Validate at even higher Reynolds numbers:** The Erturk 2005 data within the example investigates this case up to a Reynolds number of 20000.  It is an interesting exercise to simulate these more complex cases using the adjoint time-stepping ``steady_bdf`` scheme. 
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:198]:
```suggestion
Base case results (Re=400)
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:220]:
```suggestion
Higher-Reynolds case results (Re=7500)
```

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (lpsaavedra) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:265]: I would delete this and keep only one subsection for the higher Reynolds case

**Review by lpsaavedra — CHANGES_REQUESTED** [review]
Very nice! I really like how the examples are going to look from now on. I have found some typos and I have some suggestions to improve it.

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (blaisb) [reply doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:265]: You are right, it detracts the focus a bit. I'll remove that part. We can leave it to another case.

**Comment by blaisb** [comment]
All your comments have been applied @lpsaavedra . Thanks for all these nice findings :)
I removed the intiial conditions of the higher Reynolds number case.

**Review by lpsaavedra — APPROVED** [review]

**Comment by blaisb** [comment]
If everybody agrees, i'll merge this once the CI has passed.

File: examples/incompressible_flow/2d_lid_driven_cavity/Reynolds_7500/post_process_Reynolds_7500.py
Comment (oguevremont) [inline examples/incompressible_flow/2d_lid_driven_cavity/Reynolds_7500/post_process_Reynolds_7500.py:4]: I would add "and Erturk et al."

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (oguevremont) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:206]: I would replace "de" by "the"

File: doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst
Comment (oguevremont) [inline doc/source/examples/incompressible-flow/2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow.rst:243]: In the 4th sentence, "stop-tolerance" should be replaced by "stop tolerance"

**Review by oguevremont — COMMENTED** [review]
It is nice to see this example with the new documentation format. I found only a few mistakes which I could fix or someone else.

**Pull request merged**
